### PR TITLE
refacor(indexer): Add local cache for delayed local receipts. Metrics, log clean up and improvements

### PR DESCRIPTION
**Pull request by khorolets**
## Rational

Today on testnet, we encountered very weird behavior from the indexers. Indexing BPS dropped to 0.004 blocks per second. This speed is unacceptable, especially when indexing the blocks already synced by the node.

After an investigation, we found the suspect. The most obvious suspect was the process of looking for the local delayed receipt. This kind of receipt is treated differently by the indexer. On the indexer side, we tend to include them in the `StreamerMessage` while `nearcore` avoids storing them to save some space.

During some special periods (e.g., congestion), the number of local delayed receipts grows, leading to extra work by the indexer.

This PR introduces a local (in-memory) cache for the indexer instance to keep such receipts until they are needed. This should prevent the example from having to walk back through history.

When we miss the receipt, we try to find it in previous blocks. Not knowing where to look exactly, we go back block by block until we find it. However, we limit this walk up to 1000 blocks because otherwise, the indexer node would be stuck forever. By the way, that's exactly what it looks like.

## TL;DR

* Add local cache to keep local receipts if we don't observe `ExecutionOutcome` for them in the same block (Local Delayed Receipts)
* Add a metric to keep track of how far in history we go back when we do
* Clean up and improve some debug messages. I was stupid and didn't know what I was doing when I coded them three years ago 🤦‍♂️
* Add extra debug message when they were missing

### CODE REVIEW

File: chain/indexer/src/streamer/mod.rs [hunk 1]
```diff
@@ -63,7 +73,6 @@ fn test_problematic_blocks_hash() {
 /// This function supposed to return the entire `StreamerMessage`.
 /// It fetches the block and all related parts (chunks, outcomes, state changes etc.)
 /// and returns everything together in one struct
-#[async_recursion]
```
Comment (khorolets) [inline chain/indexer/src/streamer/mod.rs:66]: Please, don't ask me why it was here I don't remember and can come up with the answer

File: chain/indexer/src/streamer/mod.rs [hunk 2]
```diff
@@ -17,17 +17,27 @@ use near_indexer_primitives::{
...
+use crate::{AwaitForNodeSyncedEnum, IndexerConfig};
 
 mod errors;
 mod fetchers;
 mod metrics;
 mod utils;
 
+lazy_static! {
+    static ref DELAYED_LOCAL_RECEIPTS_CACHE: Arc<RwLock<HashMap<CryptoHash, views::ReceiptView>>> =
+        Arc::new(RwLock::new(HashMap::new()));
```
Comment (khorolets) [inline chain/indexer/src/streamer/mod.rs:38]: I am unsure about reasoning to have `Arc` here, please advise me if it's not needed here

File: chain/indexer/src/streamer/mod.rs [hunk 3]
```diff
@@ -138,6 +147,11 @@ pub async fn build_streamer_message(
             {
                 debug_assert!(outcome.receipt.is_none());
                 outcome.receipt = Some(receipt.clone());
+            } else {
+                DELAYED_LOCAL_RECEIPTS_CACHE
```
Comment (khorolets) [inline chain/indexer/src/streamer/mod.rs:151]: I want to prove the concept of having a cache. I am considering using the RocksDB instance we already use to keep track of the `last_synced_block_height` by the indexer. 🤔 Now I think the word `synced` should be replaced with `indexed`

File: chain/indexer/src/streamer/mod.rs [hunk 4]
```diff
@@ -149,34 +163,65 @@ pub async fn build_streamer_message(
...
+                        )
+                        .await?
+                        {
+                            tracing::debug!(
+                                target: INDEXER,
+                                "Found receipt {} in previous block {}",
+                                execution_outcome.id,
+                                prev_block_tried,
+                            );
+                            metrics::LOCAL_RECEIPT_LOOKUP_IN_HISTORY_BLOCKS_BACK
```
Comment (khorolets) [inline chain/indexer/src/streamer/mod.rs:218]: This metric in Grafana will allow us to see the chart of how far we went in history if we had to find the receipt.

File: chain/indexer/src/streamer/mod.rs [hunk 5]
```diff
@@ -353,9 +401,9 @@ pub(crate) async fn start(
 
                 match response {
                     Ok(streamer_message) => {
-                        debug!(target: INDEXER, "{:#?}", &streamer_message);
```
Comment (khorolets) [inline chain/indexer/src/streamer/mod.rs:356]: Trust me, the entire `StreamerMessage` in the logs does not help to debug the indexer.

File: chain/indexer/Cargo.toml [hunk 6]
```diff
@@ -22,6 +22,7 @@ serde.workspace = true
 serde_json.workspace = true
 tokio.workspace = true
 tracing.workspace = true
+lazy_static.workspace = true
```
Comment (frol) [inline chain/indexer/Cargo.toml:25]: Consider using [`std::sync::OnceLock`](https://doc.rust-lang.org/stable/std/sync/struct.OnceLock.html) instead

File: chain/indexer/src/streamer/mod.rs [hunk 7]
```diff
@@ -296,8 +341,11 @@ pub(crate) async fn start(
             .path()
             .join("indexer");
 
-    // TODO: implement proper error handling
-    let db = DB::open_default(indexer_db_path).unwrap();
+    let db = match DB::open_default(indexer_db_path) {
+        Ok(db) => db,
+        Err(err) => panic!("Unable to open indexer db: {:?}", err),
+    };
```
Comment (frol) [inline chain/indexer/src/streamer/mod.rs:347]:
```suggestion
    let db = DB::open_default(indexer_db_path).expect("Unable to open indexer db");
```

File: chain/indexer/src/streamer/mod.rs [hunk 8]
```diff
@@ -149,34 +163,65 @@ pub async fn build_streamer_message(
             let receipt = if let Some(receipt) = receipt {
                 receipt
             } else {
```
Comment (frol) [inline chain/indexer/src/streamer/mod.rs:165]: nit: [let-else](https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html) is nice

```suggestion
            let Some(receipt) = receipt else {
```

File: chain/indexer/src/streamer/mod.rs [hunk 9]
```diff
@@ -149,34 +163,65 @@ pub async fn build_streamer_message(
...
-                    };
-
-                    prev_block_hash = prev_block.header.prev_hash;
-
-                    if let Some(receipt) = find_local_receipt_by_id_in_block(
-                        &client,
-                        &runtime_config,
-                        prev_block,
+                // Attempt to find the receipt in the cache
+                if let Some(receipt) = DELAYED_LOCAL_RECEIPTS_CACHE
```
Comment (frol) [inline chain/indexer/src/streamer/mod.rs:167]: nit: let-else

**Review by frol — APPROVED** [review]
Looks good to me 👍

File: chain/indexer/src/streamer/mod.rs (hunk 2 above)
Comment (frol) [reply chain/indexer/src/streamer/mod.rs:38]: You don't need it here, you don't clone it anywhere

File: chain/indexer/src/streamer/mod.rs (hunk 7 above)
Comment (khorolets) [reply chain/indexer/src/streamer/mod.rs:347]: Will it include the original error message? I wanted it to be shown.

Comment (frol) [reply chain/indexer/src/streamer/mod.rs:347]: Yes, it will. That is the whole point of `.expect`

**Review by akhi3030 — APPROVED** [review]
approving to unblock
